junit-vintage-engine/src/main/java/org/junit/vintage/engine/VintageTestEngine.java

```diff
@@ -75,11 +88,64 @@ public void execute(ExecutionRequest request) {
 
 	private void executeAllChildren(VintageEngineDescriptor engineDescriptor,
 			EngineExecutionListener engineExecutionListener) {
+		boolean parallelExecutionEnabled = getParallelExecutionEnabled();
+
+		if (parallelExecutionEnabled) {
+			executeInParallel(engineDescriptor, engineExecutionListener);
+		}
+		else {
+			executeSequentially(engineDescriptor, engineExecutionListener);
+		}
+	}
+
+	private void executeInParallel(VintageEngineDescriptor engineDescriptor,
+			EngineExecutionListener engineExecutionListener) {
+		ExecutorService executorService = Executors.newFixedThreadPool(getThreadPoolSize());
+		RunnerExecutor runnerExecutor = new RunnerExecutor(engineExecutionListener);
+
+		List<TestDescriptor> children = new CopyOnWriteArrayList<>(engineDescriptor.getModifiableChildren());
+
+		CompletableFuture<Void> allOf = CompletableFuture.allOf(
+			children.stream().map(descriptor -> CompletableFuture.runAsync(() -> {
+				RunnerTestDescriptor testDescriptor = (RunnerTestDescriptor) descriptor;
+				try {
+					runnerExecutor.execute(testDescriptor);
+				}
+				catch (Exception e) {
+					engineExecutionListener.executionSkipped(testDescriptor, e.getMessage());
+				}
+			}, executorService)).toArray(CompletableFuture[]::new));
+
+		try {
+			allOf.get(SHUTDOWN_TIMEOUT_SECONDS, TimeUnit.SECONDS);
```

Comment:
Doesn't this limit the overall test execution duration to 60 seconds? I don't think we can make that assumption. Instead we need to wait without a timeout.

Reply:
I didn't think the test could be longer than 60 seconds... I will make the change!